extractor/extract.py

```diff
@@ -110,9 +113,41 @@ def main():
     for j in blacklist_i:
         solver.Add(x[j] == 0)
 
+    # fusion shenanigans
+    can_fuse_node_of_class = {}
+    for j in range(num_classes):
+        can_fuse_node_of_class[j] = solver.IntVar(0, 1, 'can_fuse_node_of_class[%i]' % j)
+
+    for i in range(num_nodes):
+        if not fus[i]:
+            # a class is fusable only if no un-fusable nodes are chosen
+            solver.Add(can_fuse_node_of_class[g[i]] <= 1 - x[i])
+    
+    can_fuse_node_with_children = {}
+    for i in range(num_nodes):
+        can_fuse_node_with_children[i] = solver.IntVar(0, len(h[i]), 'can_fuse_node_with_children[%i]' % j)
+        if not fus[i]:
+            # if the node itself cannot fuse, it can never fuse
+            solver.Add(can_fuse_node_with_children[i] <= 0)
+        # count children that can fuse
+        solver.Add(can_fuse_node_with_children[i] <= sum(can_fuse_node_of_class[m] for m in h[i]))
+
+    fuse_ratio = {}
```

Comment:
nit: maybe this should be `fuse_count` (or something that doesn't imply it shouldn't be an integer)